kotsadm/pkg/automation/automation.go

```diff
@@ -84,48 +92,150 @@ func AutomateInstall() error {
 		// check license expiration
 		expired, err := kotspull.LicenseIsExpired(verifiedLicense)
 		if err != nil {
-			logger.Error(err)
+			logger.Error(errors.Wrap(err, "failed to check is lincense is expired"))
 			continue
 		}
 		if expired {
-			logger.Error(errors.New("license is expired"))
+			logger.Error(errors.Errorf("license is expired for app %s", verifiedLicense.Spec.AppSlug))
 			continue
 		}
 
 		skipImagePush, err := kotsutil.IsImagesPushedSet(kotsadmtypes.KotsadmConfigMap)
 		if err != nil {
-			return errors.Wrap(err, "failed to get existing kotsadm config map")
+			logger.Error(errors.Wrap(err, "failed to get existing kotsadm config map"))
+			continue
 		}
 
 		desiredAppName := strings.Replace(verifiedLicense.Spec.AppSlug, "-", " ", 0)
 		upstreamURI := fmt.Sprintf("replicated://%s", verifiedLicense.Spec.AppSlug)
 
 		a, err := store.GetStore().CreateApp(desiredAppName, upstreamURI, string(license), verifiedLicense.Spec.IsAirgapSupported, skipImagePush)
 		if err != nil {
-			logger.Error(err)
+			logger.Error(errors.Wrap(err, "failed to create app record"))
 			continue
 		}
 
-		// complete the install online
-		pendingApp := installationtypes.PendingApp{
-			ID:          a.ID,
-			Slug:        a.Slug,
-			Name:        a.Name,
-			LicenseData: string(license),
-		}
-		_, err = online.CreateAppFromOnline(&pendingApp, upstreamURI, true)
+		airgapData, err := getAirgapData(clientset, verifiedLicense)
 		if err != nil {
-			logger.Error(err)
+			logger.Error(errors.Wrapf(err, "failed to load airgap data for %s", verifiedLicense.Spec.AppSlug))
 			continue
 		}
 
+		if skipImagePush && airgapData != nil {
+			airgapFilesDir, err := ioutil.TempDir("", "headless-airgap")
+			if err != nil {
+				logger.Error(errors.Wrap(err, "failed to create temp dir"))
+				continue
+			}
+			defer os.RemoveAll(airgapFilesDir)
+
+			for filename, data := range airgapData {
+				err := ioutil.WriteFile(filepath.Join(airgapFilesDir, filename), data, 0644)
+				if err != nil {
+					logger.Error(errors.Wrapf(err, "failed to create file %s", filename))
+					continue LICENSE_LOOP
+				}
+			}
+
+			kotsadmOpts, err := kotsadm.GetKotsadmOptionsFromCluster(os.Getenv("POD_NAMESPACE"), clientset)
+			if err != nil {
+				logger.Error(errors.Wrap(err, "failed to load registry info"))
+				continue
+			}
+
+			registryHost := kotsadmOpts.OverrideRegistry
+			namespace := kotsadmOpts.OverrideNamespace
+			username := kotsadmOpts.Username
+			password := kotsadmOpts.Password
+
+			pendingApp := airgaptypes.PendingApp{
+				ID:          a.ID,
+				Slug:        a.Slug,
+				Name:        a.Name,
+				LicenseData: string(license),
+			}
+
+			err = airgap.CreateAppFromAirgap(&pendingApp, airgapFilesDir, registryHost, namespace, username, password, true)
+			if err != nil {
+				logger.Error(errors.Wrap(err, "failed to create airgap app"))
+				continue
+			}
+		} else {
+			pendingApp := onlinetypes.PendingApp{
```

Comment:
same here, i've lost context of the else.  can you explain more about the use case this is handling (in a comment?)